compiler/noirc_evaluator/src/ssa/opt/remove_unreachable_instructions.rs

```diff
@@ -0,0 +1,430 @@
+        "#);
+    }
+
+    #[test]
+    fn removes_unreachable_instructions_from_successors_goes_in_pre_order() {
+        let src = r#"
+        acir(inline) predicate_pure fn main f0 {
+          b0():
+            v0 = make_array [] : [&mut u1; 0]
+            constrain u1 0 == u1 1, "Index out of bounds"
+            v4 = array_get v0, index u32 0 -> &mut u1
+            v5 = load v4 -> u1
+            jmp b2(v5)
+          b1(v8: u1):
+            v9 = add v8, u1 1
+            return v9
+          b2(v6: u1):
+            v7 = add v6, u1 1
+            jmp b1(v7)
+        }
+        "#;
+        let ssa = Ssa::from_str(src).unwrap();
+        let ssa = ssa.remove_unreachable_instructions();
+
+        assert_ssa_snapshot!(ssa, @r#"
+        acir(inline) predicate_pure fn main f0 {
+          b0():
+            v2 = make_array [] : [&mut u1; 0]
+            constrain u1 0 == u1 1, "Index out of bounds"
+            jmp b2(u1 0)
+          b1(v0: u1):
+            return u1 0
+          b2(v1: u1):
+            jmp b1(u1 0)
+        }
+        "#);
+    }
+
+    #[test]
+    fn removes_unreachable_instructions_if_successor_has_other_predecessors() {
+        let src = r#"
+        acir(inline) predicate_pure fn main f0 {
+          b0():
+            constrain u1 0 == u1 1, "Index out of bounds"
+            jmp b1()
+          b1():
+            v1 = add Field 1, Field 2
+            jmpif u1 0 then: b2, else: b3
+          b2():
+            v2 = add Field 1, Field 2
+            jmp b1()
+          b3():
+            v3 = add Field 1, Field 2
+            return v3
+        }
+        "#;
+        let ssa = Ssa::from_str(src).unwrap();
+        let ssa = ssa.remove_unreachable_instructions();
+
+        assert_ssa_snapshot!(ssa, @r#"
+        acir(inline) predicate_pure fn main f0 {
+          b0():
+            constrain u1 0 == u1 1, "Index out of bounds"
+            jmp b1()
+          b1():
+            jmpif u1 0 then: b2, else: b3
+          b2():
+            jmp b1()
+          b3():
+            return Field 0
+        }
+        "#);
+    }
+
+    #[test]
+    fn removes_unreachable_instructions_following_block_with_no_instructions() {
+        let src = r#"
+        acir(inline) predicate_pure fn main f0 {
+          b0():
+            constrain u1 0 == u1 1, "Index out of bounds"
+            jmp b1()
+          b1():
+            jmp b2()
+          b2():
+            v2 = add Field 1, Field 2
+            return v2
+        }
+        "#;
+        let ssa = Ssa::from_str(src).unwrap();
+        let ssa = ssa.remove_unreachable_instructions();
+
+        assert_ssa_snapshot!(ssa, @r#"
+        acir(inline) predicate_pure fn main f0 {
+          b0():
+            constrain u1 0 == u1 1, "Index out of bounds"
+            jmp b1()
+          b1():
+            jmp b2()
+          b2():
+            return Field 0
+        }
+        "#);
+    }
+
+    #[test]
+    fn does_not_zeroes_terminator_of_previously_seen_blocks() {
```

Comment:
```suggestion
    fn does_not_zeroes_terminator_of_non_dominated_blocks() {
```

I commented on this but it got lost: I think it "previously seen" assumes we are much familiar with the traversal algorithm, rather than stating an invariant. For example here I think we will visit blocks in the order of `b0, b1, b2, b3, b4`; I guess the test name highlights that we won't visit `b1` again because we have already seen it, and therefore it keeps its `add`, but `b4` also keeps the `add`, despite being visited after `b3`. The real reason both get to keep their instructions is they are both not dominated by `b3`. 

I realise this must be one of the tests you created earlier, just wanted to point it out that it could be renamed.